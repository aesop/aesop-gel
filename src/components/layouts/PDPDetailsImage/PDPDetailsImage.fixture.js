export default {
  image: {
    altText: 'here is a descriptive alt tag',
    sizes: [
      {
        viewport: 'LARGE',
        src: '/images/products/Skin_Amazing_Face_Cleanser_100mL_large.png 2x',
      },
      {
        viewport: 'MEDIUM',
        src: '/images/products/Skin_Amazing_Face_Cleanser_100mL_medium.png 2x',
      },
      {
        viewport: 'SMALL',
        src: '/images/products/Skin_Amazing_Face_Cleanser_100mL_small.png 2x',
      },
    ],
  },
};
