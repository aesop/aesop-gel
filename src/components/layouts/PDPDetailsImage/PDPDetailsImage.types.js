import { ChangeEvent } from 'react';
import { ClassSetType } from '~/types';
import { ProductVariantImage } from '~/types/Product';

interface VariantOptions {
  sku: string;
  size: string | undefined;
}

export interface Props extends ClassSetType {
  image?: ProductVariantImage | undefined;
  variantOptions: VariantOptions[];
  id?: string | undefined;
  onVariantChange: (event: ChangeEvent<HTMLInputElement>) => void;
  selectedOption: string;
}

export { Props as default }; // eslint-disable-line no-undef
