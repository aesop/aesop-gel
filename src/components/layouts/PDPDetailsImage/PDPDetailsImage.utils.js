interface Size {
  src: string;
  viewport: string;
}

export const imageSource = (
  viewport: string,
  sizes: Size[],
): string | undefined => {
  const viewportRegex = new RegExp(viewport, 'i');
  const viewportImage = sizes.find(size => viewportRegex.test(size.viewport));

  if (!viewportImage) {
    return undefined;
  }

  return `${process.env.GATSBY_MEDIA || ''}${viewportImage.src}`;
};

export default {
  imageSource,
};
