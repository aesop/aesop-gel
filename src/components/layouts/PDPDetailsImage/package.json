{
  "name": "PDPDetailsImage",
  "version": "0.0.0",
  "main": "PDPDetailsImage.js"
}
