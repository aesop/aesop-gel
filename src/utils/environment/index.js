/* istanbul ignore file */
import { isInBrowser } from './environment';

export { isInBrowser };

export default { isInBrowser };
