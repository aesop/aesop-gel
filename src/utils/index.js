export * as environment from './environment';
export * as objects from './objects';
export * as product from './product';
export * as viewports from './viewports';
