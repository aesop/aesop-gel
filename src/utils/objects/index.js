import { isObjectPopulatedArray } from './objects';

export { isObjectPopulatedArray };

export default { isObjectPopulatedArray };
