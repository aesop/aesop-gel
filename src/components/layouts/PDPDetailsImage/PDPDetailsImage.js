import React from 'react';
import cx from 'classnames';
import { BREAKPOINTS } from '~/constants';
import Image from '~/components/elements/Image';
import RadioGroup from '~/components/controls/RadioGroup';
import Transition from '~/components/elements/Transition';
import Props from './PDPDetailsImage.types';
import { imageSource } from './PDPDetailsImage.utils';
import styles from './PDPDetailsImage.module.css';

const { LARGE, MEDIUM, SMALL } = BREAKPOINTS;

const PDPDetailsImage: React.FunctionComponent<Props> = ({
  className,
  id,
  image,
  variantOptions,
  onVariantChange,
  selectedOption,
}) => {
  if (!image) {
    return null;
  }

  const classSet: string = cx(styles.base, className);
  const { altText, sizes } = image;

  const variantRadioOptions = variantOptions.map(option => ({
    label: option.size || '',
    value: option.sku,
  }));

  return (
    <Transition activeOnMount={true} type="shiftInLeft">
      <div className={classSet} id={`product-${id}`}>
        <Image
          altText={altText}
          className={styles.image}
          large={imageSource(LARGE.VIEWPORT, sizes)}
          medium={imageSource(MEDIUM.VIEWPORT, sizes)}
          small={imageSource(SMALL.VIEWPORT, sizes)}
        />
        {variantOptions.length === 1 ? (
          <div className={styles.singleVariantLabel}>
            {variantOptions[0].size}
          </div>
        ) : (
          <RadioGroup
            className={styles.variants}
            name="sku"
            onChange={e => onVariantChange(e)}
            options={variantRadioOptions}
            testReference="selectedVariant"
            value={selectedOption}
          />
        )}
      </div>
    </Transition>
  );
};

export default PDPDetailsImage;
