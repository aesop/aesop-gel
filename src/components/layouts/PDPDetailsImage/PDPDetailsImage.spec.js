import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer, { act } from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';
import PDPDetailsImage from './PDPDetailsImage';
import { imageSource } from './PDPDetailsImage.utils';

configure({ adapter: new Adapter() });

const mockFn = jest.fn();

const image = {
  altText: 'here is a descriptive alt tag',
  sizes: [
    {
      viewport: 'LARGE',
      src: '/images/products/Skin_Amazing_Face_Cleanser_100mL_large.png 2x',
    },
    {
      viewport: 'MEDIUM',
      src: '/images/products/Skin_Amazing_Face_Cleanser_100mL_medium.png 2x',
    },
    {
      viewport: 'SMALL',
      src: '/images/products/Skin_Amazing_Face_Cleanser_100mL_small.png 2x',
    },
  ],
};

describe('<PDPDetailsImage />', () => {
  it('should be defined', () => {
    expect(PDPDetailsImage).toBeDefined();
  });

  it('renders base component correctly', async () => {
    const component = renderer.create(
      <PDPDetailsImage
        id="PDPDetailsImage"
        image={image}
        onVariantChange={mockFn}
        selectedOption="aesop"
        variantOptions={[{ size: '100ml', sku: 'aesop' }]}
      />,
    );

    await act(async () => {
      await waitForExpect(() => {
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  it('should return `null` if no image is passed', () => {
    const component = shallow(
      <PDPDetailsImage
        id="PDPDetailsImage"
        onVariantChange={mockFn}
        selectedOption="aesop"
        variantOptions={[{ size: '100ml', sku: 'aesop' }]}
      />,
    );
    expect(component.type()).toEqual(null);
  });
});

describe('PDPDetailsImage.utils.imageSource', () => {
  it('should return undefined if viewportImage is not found', () => {
    const actual = imageSource('LARGE.VIEWPORT', [
      {
        src: 'foo',
        viewport: 'bar',
      },
    ]);

    expect(actual).toBe(undefined);
  });

  it('should return the src of a found viewport image set', () => {
    const actual = imageSource('LARGE', image.sizes);

    expect(actual).toBe(
      `${process.env.GATSBY_MEDIA}/images/products/Skin_Amazing_Face_Cleanser_100mL_large.png 2x`,
    );
  });
});

describe('PDPDetailsImage.utils.imageSource env vars', () => {
  const CURRENT_ENV = process.env;

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...CURRENT_ENV };
    delete process.env.GATSBY_MEDIA;
  });

  afterEach(() => {
    process.env = CURRENT_ENV;
  });

  it('should return a correct image url even if the process.env.media is set to falsy', () => {
    process.env.GATSBY_MEDIA = '';

    const actual = imageSource('LARGE', image.sizes);

    expect(actual).toBe(
      `${process.env.GATSBY_MEDIA}/images/products/Skin_Amazing_Face_Cleanser_100mL_large.png 2x`,
    );
  });
});
