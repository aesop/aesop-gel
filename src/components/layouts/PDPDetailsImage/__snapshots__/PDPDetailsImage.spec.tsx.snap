// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PDPDetailsImage /> renders base component correctly 1`] = `
<div
  className="base"
  id="product-PDPDetailsImage"
>
  <picture
    className="picture base image"
  >
    <source
      media="(min-width: 1025px)"
      srcSet="https://acceptance.aesop.com/images/products/Skin_Amazing_Face_Cleanser_100mL_large.png 2x"
    />
    <source
      media="(min-width: 640px)"
      srcSet="https://acceptance.aesop.com/images/products/Skin_Amazing_Face_Cleanser_100mL_medium.png 2x"
    />
    <source
      media="(min-width: 0px)"
      srcSet="https://acceptance.aesop.com/images/products/Skin_Amazing_Face_Cleanser_100mL_small.png 2x"
    />
    <img
      alt="here is a descriptive alt tag"
    />
  </picture>
  <div
    className="singleVariantLabel"
  >
    100ml
  </div>
</div>
`;
